File: CompleteBinaryTrees/complete_binary_tree.js
const Node = require("../Nodes/node"); // Import the Node class to represent each node in the tree

/**
 * Class representing a Complete Binary Tree.
 * A Complete Binary Tree is a binary tree where all levels are completely filled
 * except possibly for the last level, which is filled from left to right.
 */
class CompleteBinaryTree {
    constructor() {
        this.nodes = [];  // Array to store nodes in level-order, ensuring completeness
    }

    /**
     * Inserts a new node in the complete binary tree.
     * @param {any} data - The value to be stored in the new node.
     */
    insert(data) {
        const new_node = new Node(data); // Create a new Node instance
        this.nodes.push(new_node);  // Add the new node to the array

        // Calculate the parent index based on the new node's position in the array
        const index = this.nodes.length - 1;  // Index of the newly added node
        if (index > 0) {  // Check if the node is not the root
            const parentIndex = Math.floor((index - 1) / 2);  // Calculate the parent index
            const parent = this.nodes[parentIndex];  // Get the parent node

            // Assign the new node as the left or right child of its parent
            if (!parent.left) {  // If the left child is not assigned
                parent.left = new_node;  // Assign the new node as the left child
            } else {
                parent.right = new_node;  // Otherwise, assign it as the right child
            }
        }
    }

    /**
     * Searches for a target value in the tree.
     * @param {any} target - The value to search for in the tree.
     * @returns {boolean} True if the target is found, otherwise false.
     */
    search(target) {
        // Iterate through the nodes array to find the target value
        for (const node of this.nodes) {
            if (node.data === target) {  // Compare node data with the target
                return true;  // Return true if found
            }
        }
        return false;  // Return false if the target is not found
    }

    /**
     * Prints the tree in level-order.
     * Level-order traversal prints nodes from top to bottom, left to right.
     */
    printTree() {
        if (this.nodes.length === 0) return;  // If there are no nodes, do nothing

        const queue = [this.nodes[0]];  // Start with the root node in the queue
        const values = [];  // Array to collect node values for printing

        while (queue.length > 0) {  // Process nodes until the queue is empty
            const currentNode = queue.shift();  // Remove the first node from the queue
            if (currentNode) {
                values.push(currentNode.data);  // Store the current node's data

                // Queue the left and right children if they exist
                if (currentNode.left !== null) queue.push(currentNode.left);
                if (currentNode.right !== null) queue.push(currentNode.right);
            }
        }

        console.log(values.join(" "));  // Print all collected values in a single line
    }
}

module.exports = CompleteBinaryTree;  // Export the CompleteBinaryTree class for use in other modules

File: CompleteBinaryTrees/complete_binary_tree.py
from Nodes.node import Node  # Import the Node class to represent each node in the tree

class CompleteBinaryTree:
    def __init__(self):
        """Initializes a CompleteBinaryTree instance.

        The tree is represented using a list to store nodes in level order.
        """
        self.nodes = []  # Using a list to store nodes in level order

    def insert(self, data):
        """Inserts a new node into the Complete Binary Tree while maintaining the complete tree property.

        Nodes are added from left to right.

        Args:
            data (any): The value to be stored in the new node.
        """
        new_node = Node(data)  # Create a new Node instance with the provided data
        self.nodes.append(new_node)  # Add the new node to the list

        # If not the root, determine the parent and add the new node as a child
        idx = len(self.nodes) - 1  # Get the index of the newly added node
        if idx > 0:  # Check if the node is not the root
            parent_idx = (idx - 1) // 2  # Calculate the parent index
            # Assign the new node as the left or right child of its parent
            if not self.nodes[parent_idx].left:  # If the left child is not assigned
                self.nodes[parent_idx].left = new_node  # Assign as left child
            else:
                self.nodes[parent_idx].right = new_node  # Otherwise, assign as right child

    def search(self, target):
        """Searches for a target value in the Complete Binary Tree.

        Args:
            target (any): The value to search for in the tree.

        Returns:
            bool: True if the target is found; otherwise, False.
        """
        for node in self.nodes:  # Iterate through the list of nodes
            if node.data == target:  # Compare the current node's data with the target
                return True  # Return True if found
        return False  # Return False if the target is not found

    def print_tree(self):
        """Prints the values of the Complete Binary Tree in level order."""
        for node in self.nodes:  # Iterate through the list of nodes
            print(node.data, end=" ")  # Print each node's data in a single line
